core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedNodeRoutingTests.java

```diff
@@ -91,4 +120,131 @@ public void testSimpleFailedNodeTest() {
+                .add(createNode()).build();
+            state = ClusterState.builder(state).nodes(newNodes).build();
+            state = cluster.reroute(state, new ClusterRerouteRequest()); // always reroute after node leave
+        }
+
+        // Log the shard versions (for debugging if necessary)
+        for (ObjectCursor<DiscoveryNode> cursor : state.nodes().getDataNodes().values()) {
+            Version nodeVer = cursor.value.getVersion();
+            logger.info("--> node [{}] has version [{}]", cursor.value.getId(), nodeVer);
+        }
+
+        // randomly create some indices
+        logger.info("--> creating some indices");
+        for (int i = 0; i < randomIntBetween(2, 5); i++) {
+            String name = "index_" + randomAlphaOfLength(8).toLowerCase(Locale.ROOT);
+            Settings.Builder settingsBuilder = Settings.builder()
+                .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 4))
+                .put(SETTING_NUMBER_OF_REPLICAS, randomIntBetween(2, 4));
+            CreateIndexRequest request = new CreateIndexRequest(name, settingsBuilder.build()).waitForActiveShards(ActiveShardCount.NONE);
+            state = cluster.createIndex(state, request);
+            assertTrue(state.metaData().hasIndex(name));
+        }
+        state = cluster.reroute(state, new ClusterRerouteRequest());
+
+        ClusterState previousState = state;
+
+        logger.info("--> starting shards");
+        state = cluster.applyStartedShards(state, state.getRoutingNodes().shardsWithState(INITIALIZING));
+        logger.info("--> starting replicas a random number of times");
+        for (int i = 0; i < randomIntBetween(1,10); i++) {
+            state = cluster.applyStartedShards(state, state.getRoutingNodes().shardsWithState(INITIALIZING));
+        }
+
+        logger.info("--> state before failing shards: {}", state);
+        for (ShardRouting shardRouting : state.getRoutingNodes().shardsWithState(STARTED)) {
+            if (shardRouting.primary() && randomBoolean()) {
+                ShardRouting replicaToBePromoted = state.getRoutingNodes()
+                    .activeReplicaWithHighestVersion(shardRouting.shardId());
+                final ClusterState currentState = state;
+                // List of potential candidate replicas for promotion
+                Set<ShardRouting> candidates = state.getRoutingNodes().shardsWithState(STARTED)
+                    .stream()
+                    .filter(s -> !s.primary() && s.active())
+                    .filter(s -> s.shardId().equals(shardRouting.shardId()))
+                    .filter(s -> !s.equals(replicaToBePromoted))
+                    .filter(s -> currentState.getRoutingNodes().node(s.currentNodeId()) != null)
+                    .collect(Collectors.toSet());
+                // If we find a replica and at least another candidate
+                if (replicaToBePromoted != null && candidates.size() > 0) {
```

Comment:
we don't need to determine `replicaToBePromoted`. Candidates also does not need to be filtered with ` !s.equals(replicaToBePromoted)`. It's ok to just check candidates.size() > 0 here to see whether there is going to be a new primary. In that case, we fail the primary + `random(0, candidates.size() - 1)` replicas and check afterwards that the new primary is on a node that is at least as high as all replicas.